### PR TITLE
Opcode rol

### PR DESCRIPTION
**Pull request by sa12123**
Added opcode rol, addressing modes abs and absx

### CODE REVIEW

File: core/src/mos6502.cpp [hunk 1]
```diff
@@ -609,6 +613,25 @@ Pipeline Mos6502::create_eor_instruction(Opcode opcode) {
     return result;
 }
 
+Pipeline Mos6502::create_rol_instruction(Opcode opcode) {
+    const MemoryAccess memory_access = get_memory_access(opcode.family);
+    Pipeline result;
+    result.append(create_addressing_steps(opcode.address_mode, memory_access));
```
Comment (johnor) [inline core/src/mos6502.cpp]: Rol-instructions are read-modify-write according to: http://atarihq.com/danb/files/64doc.txt
We will have to add support for those if needed in create_addressing_steps and create_absolute_indexed_addressing_steps.

File: core/src/mos6502.cpp [hunk 2]
```diff
@@ -609,6 +613,25 @@ Pipeline Mos6502::create_eor_instruction(Opcode opcode) {
...
+
+    result.push([=]() {
+        const uint8_t operand = mmu_->read_byte(effective_address_);
+        const uint8_t carry = registers_->p & C_FLAG ? 0x01 : 0x00;
+        const uint16_t temp_result = (operand << 1) | carry;
+        registers_->a = static_cast<uint8_t>(temp_result);
```
Comment (johnor) [inline core/src/mos6502.cpp]: Looks like only the Accumulator addressing mode for ROL should modify a. The rest should read from memory, and write back to the same memory address.

File: core/src/opcode.cpp [hunk 3]
```diff
@@ -254,6 +258,7 @@ MemoryAccess get_memory_access(const Family family) {
     case Family::DEY:
     case Family::DEX:
     case Family::EOR:
+    case Family::ROL:
```
Comment (johnor) [inline core/src/opcode.cpp]: Should be moved to ReadWrite

File: core/test/src/test_cpu.cpp [hunk 4]
```diff
@@ -2480,4 +2482,94 @@ TEST_F(CpuTest, eor_absy_without_page_crossing) {
...
 
+// ROL, ABS
+TEST_F(CpuAbsoluteTest, rol_abs_set_neg_clear_zero) {
+    registers.a = 0b00000000;
+    registers.p = Z_FLAG;
+    expected.a = 0b10101010;
+    expected.p = N_FLAG;
+    memory_content = 0b01010101;
+
+    run_read_instruction(ROL_ABS);
```
Comment (johnor) [inline core/test/src/test_cpu.cpp]: rol absolute should take 6 cycles. Need to add a run_read_write_instruction member to the test fixture and add support for read write instructions there

File: core/test/src/test_cpu.cpp [hunk 5]
```diff
@@ -2480,4 +2482,94 @@ TEST_F(CpuTest, eor_absy_without_page_crossing) {
...
+    stage_instruction(ROL_ABSX);
+
+    expected.pc += 2;
+    expected.a = 0b01100110;
+    expected.p = 0;
+
+    ON_CALL(mmu, read_word(0x4322)).WillByDefault(Return(0x5678));
+    EXPECT_CALL(mmu, read_byte(0x5678 + 0x10)).WillOnce(Return(0b00110011));
+
+    step_execution(4);
```
Comment (johnor) [inline core/test/src/test_cpu.cpp]: Rol abs x should take 7 cycles.

File: core/src/mos6502.cpp (hunk 1 above)
Comment (sa12123) [reply core/src/mos6502.cpp]: Ok, need to do my homework a bit more...

File: core/src/mos6502.cpp (hunk 2 above)
Comment (sa12123) [reply core/src/mos6502.cpp]: Ok, need to do my homework a bit more...

File: core/src/opcode.cpp (hunk 3 above)
Comment (sa12123) [reply core/src/opcode.cpp]: Ok, need to do my homework a bit more...

File: core/test/src/test_cpu.cpp (hunk 4 above)
Comment (sa12123) [reply core/test/src/test_cpu.cpp]: Ok, need to do my homework a bit more...

File: core/test/src/test_cpu.cpp (hunk 5 above)
Comment (sa12123) [reply core/test/src/test_cpu.cpp]: Ok, need to do my homework a bit more...

**Comment by johnor** [comment]
Replaced by #158

**Pull request closed without merging**